pkg/controller/storageversionmigrator/storageversionmigrator.go

```diff
@@ -0,0 +1,302 @@
+	if !cache.WaitForNamedCacheSync(svmc.dependencyGraphBuilder.Name(), ctx.Done(), func() bool {
+		return svmc.dependencyGraphBuilder.IsResourceSynced(gvr)
+	}) {
+		return fmt.Errorf("dependency graph is not synced, requeuing to attempt again")
+	}
+
+	gcListResourceVersion, err := convertResourceVersionToInt(svmc.dependencyGraphBuilder.GetMonitorController(gvr).LastSyncResourceVersion())
+	if err != nil {
+		return err
+	}
+	listResourceVersion, err := convertResourceVersionToInt(toBeProcessedSVM.Status.ResourceVersion)
+	if err != nil {
+		return err
+	}
+
+	if gcListResourceVersion < listResourceVersion {
+		return fmt.Errorf("GC cache is not up to date, requeuing to attempt again. gcListResourceVersion: %d, listResourceVersion: %d", gcListResourceVersion, listResourceVersion)
+	}
+
+	toBeProcessedSVM, err = svmc.kubeClient.StoragemigrationV1alpha1().
+		StorageVersionMigrations().
+		UpdateStatus(
+			ctx,
+			setStatusConditions(toBeProcessedSVM, svmv1alpha1.MigrationRunning, migrationRunningStatusReason),
+			metav1.UpdateOptions{},
+		)
+	if err != nil {
+		return err
+	}
+
+	gvk, err := svmc.restMapper.KindFor(gvr)
+	if err != nil {
+		return err
+	}
+	typeMeta := metav1.TypeMeta{}
+	typeMeta.APIVersion, typeMeta.Kind = gvk.ToAPIVersionAndKind()
+	data, err := json.Marshal(typeMeta)
+	if err != nil {
+		return err
+	}
+
+	// ToDo: implement a mechanism to resume migration from the last migrated resource in case of a failure
+	// process storage migration
+	for _, gvrKey := range svmc.dependencyGraphBuilder.GetMonitorStore(gvr).ListKeys() {
+		namespace, name, err := cache.SplitMetaNamespaceKey(gvrKey)
+		if err != nil {
+			return err
+		}
+
+		_, err = svmc.dynamicClient.Resource(gvr).
+			Namespace(namespace).
+			Patch(ctx,
+				name,
+				types.ApplyPatchType,
+				data,
+				metav1.PatchOptions{
+					FieldManager: svmc.controllerName,
+				},
+			)
+		if err != nil {
+			// in case of NotFound or Conflict, we can stop processing migration for that resource
+			if apierrors.IsNotFound(err) || apierrors.IsConflict(err) {
+				continue
+			}
+
+			_, err = svmc.kubeClient.StoragemigrationV1alpha1().
+				StorageVersionMigrations().
+				UpdateStatus(
+					ctx,
+					setStatusConditions(toBeProcessedSVM, svmv1alpha1.MigrationFailed, migrationFailedStatusReason),
+					metav1.UpdateOptions{},
+				)
+			if err != nil {
+				return err
+			}
+			logger.V(4).Error(err, "Failed to migrate the resource", "name", gvrKey, "gvr", gvr.String(), "reason", apierrors.ReasonForError(err))
+
+			return nil
+			//Todo: add retry for scenarios where API server returns rate limiting error
+		}
+		logger.V(4).Info("Successfully migrated the resource", "name", gvrKey, "gvr", gvr.String())
+	}
+
+	_, err = svmc.kubeClient.StoragemigrationV1alpha1().
+		StorageVersionMigrations().
+		UpdateStatus(
+			ctx,
+			setStatusConditions(toBeProcessedSVM, svmv1alpha1.MigrationSucceeded, migrationSuccessStatusReason),
```

Comment:
where is running set to false?

Reply:
it's handled inside `setStatusConditions`

```go
func setStatusConditions(
	toBeUpdatedSVM *svmv1alpha1.StorageVersionMigration,
	conditionType svmv1alpha1.MigrationConditionType,
	reason string,
) *svmv1alpha1.StorageVersionMigration {
	if !IsConditionTrue(toBeUpdatedSVM, conditionType) {
		if conditionType == svmv1alpha1.MigrationSucceeded || conditionType == svmv1alpha1.MigrationFailed {
			runningConditionIdx := indexOfCondition(toBeUpdatedSVM, svmv1alpha1.MigrationRunning)
			if runningConditionIdx != -1 {
				toBeUpdatedSVM.Status.Conditions[runningConditionIdx].Status = corev1.ConditionFalse
			}
		}

		toBeUpdatedSVM.Status.Conditions = append(toBeUpdatedSVM.Status.Conditions, svmv1alpha1.MigrationCondition{
			Type:           conditionType,
			Status:         corev1.ConditionTrue,
			LastUpdateTime: metav1.Now(),
			Reason:         reason,
		})
	}

	return toBeUpdatedSVM
}
```